Tools/AP_Bootloader/bl_protocol.cpp

```diff
@@ -602,11 +623,57 @@ bootloader(unsigned timeout)
             case PROTO_DEVICE_EXTF_SIZE:
                 cout((uint8_t *)&board_info.extf_size, sizeof(board_info.extf_size));
                 break;
+            
+            case PROTO_DEVICE_BL_SOFTWARE:
+            {
+                uint32_t git_hash = GIT_VERSION_INT;
+                cout((uint8_t *)&git_hash, sizeof(git_hash));
+                break;
+            }
+
+            case PROTO_DEVICE_BL_OPTS:
+            {
+                uint32_t options = 0;
+#if AP_CHECK_FIRMWARE_ENABLED
+#if AP_SIGNED_FIRMWARE
+                options |= OptionSignedFirmware;
+#else
+                options |= OptionUnsignedFirmware;
+#endif
+                if (check_good_firmware() == check_fw_result_t::CHECK_FW_OK) {
+                    options |= OptionCheckFWOk;
+                }
+#endif
+#if AP_BOOTLOADER_FLASH_FROM_SD_ENABLED
+                options |= OptionFlashFromSDCard;
+#endif
+                cout((uint8_t *)&options, sizeof(options));
+                break;
+            }
+
+            case PROTO_DEVICE_APP_SOFTWARE:
+            {
+#if AP_CHECK_FIRMWARE_ENABLED
+                uint32_t git_hash;
+                const auto *ad = get_app_descriptor();
+                
+                if (ad != nullptr) {
+                    // app_descriptor is valid, grab the hash and output
+                    git_hash = ad->git_hash;
+                    cout((uint8_t *)&git_hash, sizeof(git_hash));
```

Comment:
This probably needs to be 64 bits.

I've seen problems in our tooling which I think comes down to us now having more commits and increased chance of collisions, so a 32-bit subset of the sha yields collisions.  Sadly I haven't actually chased this problem down, so only half-remembering; the workaround is to slightly change a commit...